pmd-visualforce/src/main/java/net/sourceforge/pmd/lang/vf/rule/security/VfUnescapeElRule.java

```diff
@@ -0,0 +1,323 @@
+                        isEscaped = true;
+                        continue;
+                    }
+
+                    if (e.equals(ESCAPING.ANY)) {
+                        for (ESCAPING esc : ESCAPING.values()) {
+                            if (id.getImage().equalsIgnoreCase(esc.toString())) {
+                                isEscaped = true;
+                                break;
+                            }
+                        }
+                    }
+
+                }
+
+                if (!isEscaped) {
+                    nonEscapedIds.add(id);
+                }
+            }
+
+        }
+
+        return !nonEscapedIds.isEmpty();
+    }
+
+    private boolean containsSafeFields(final AbstractVFNode expression) {
+
+        for (int i = 0; i < expression.jjtGetNumChildren(); i++) {
+            Node child = expression.jjtGetChild(i);
+
+            if (child instanceof ASTIdentifier) {
+                if ("Id".equalsIgnoreCase(child.getImage()) || "size".equalsIgnoreCase(child.getImage())) {
+                    return true;
+                }
+            }
+
+            if (child instanceof ASTDotExpression) {
+                return containsSafeFields((ASTDotExpression) child);
+            }
+
+        }
+
+        return false;
+    }
+
+    private boolean doesTagSupportEscaping(final ASTElement node) {
+        if (node.getName() == null) {
+            return false;
+        }
+
+        switch (node.getName().toLowerCase()) { // vf is case insensitive
+        case APEX_OUTPUT_TEXT:
+        case APEX_PAGE_MESSAGE:
+        case APEX_PAGE_MESSAGES:
+        case APEX_SELECT_OPTION:
+            return true;
+        default:
+            return false;
+        }
+
+    }
+
+    private Set<ASTElExpression> hasELInInnerElements(final ASTElement node) {
+        final Set<ASTElExpression> toReturn = new HashSet<>();
+        final ASTContent content = node.getFirstChildOfType(ASTContent.class);
+        if (content != null) {
+            final List<ASTElement> innerElements = content.findChildrenOfType(ASTElement.class);
+            for (final ASTElement element : innerElements) {
+                if (element.getName().equalsIgnoreCase(APEX_PARAM)) {
+                    final List<ASTAttribute> innerAttributes = element.findChildrenOfType(ASTAttribute.class);
+                    for (ASTAttribute attrib : innerAttributes) {
+                        final List<ASTElExpression> elsInVal = attrib.findDescendantsOfType(ASTElExpression.class);
+                        for (final ASTElExpression el : elsInVal) {
+                            if (doesElContainAnyUnescapedIdentifiers(el, ESCAPING.HTMLENCODE)) {
+                                toReturn.add(el);
+                            }
+
+                        }
+                    }
+                }
+            }
+        }
+
+        return toReturn;
+    }
+
+    enum ESCAPING {
```

Comment:
@jsotuyod I had a look at the checkstyle rules. It indeed doesn't detect this case - only uppercase enum name. We use the default [TypeName](http://checkstyle.sourceforge.net/config_naming.html#TypeName) check, and the regex just ensures, that the type name begins with an uppercase letter. But after that, any combination of upper/lower case letters are possible - including all uppercase...
There is however [AbbreviationAsWordInName](http://checkstyle.sourceforge.net/config_naming.html#AbbreviationAsWordInName). This check could detect such cases. However, it will probably require  `net.sourceforge.pmd.PMD` to become `net.sourceforge.pmd.Pmd`.... which looks more correct, but would be a very drastic change....